include/revng/ADT/GenericGraph.h

```diff
@@ -166,43 +223,85 @@ class ForwardNode : public std::conditional_t<HasParent,
   NeighborContainer Successors;
 };
 
+namespace detail {
+
+/// To remove clutter from BidirectionalNode, the computation of some types are
+/// done in this class.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename,
+                  typename,
+                  bool,
+                  size_t>
+         class BidirectionalNode> // At this step BidirectionalEdge has not been
+                                  // declared yet, thus we accept a template
+                                  // parameter that has the same signature as
+                                  // BidirecationEdge that will be declared
+                                  // later. This allows us to use it as if it
+                                  // was delcared, provided that only the real
+                                  // BidirecationalEdge is used as this
+                                  // argument.
```

Comment:
These comments are too long for being inline. Put them in a `/// \note:` comment above.